pydicom/encoders/native.py

```diff
@@ -0,0 +1,170 @@
+            rle_data.extend(segment)
+            seg_lengths.append(len(segment))
+
+    # Add the number of segments to the header
+    rle_header = bytearray(pack('<L', len(seg_lengths)))
+
+    # Add the segment offsets, starting at 64 for the first segment
+    # We don't need an offset to any data at the end of the last segment
+    offsets = [64]
+    for ii, length in enumerate(seg_lengths[:-1]):
+        offsets.append(offsets[ii] + length)
+    rle_header.extend(pack('<{}L'.format(len(offsets)), *offsets))
+
+    # Add trailing padding to make up the rest of the header (if required)
+    rle_header.extend(b'\x00' * (64 - len(rle_header)))
+
+    return bytes(rle_header + rle_data)
+
+
+def _encode_segment(src: bytes, **kwargs) -> bytearray:
+    """Return `src` as an RLE encoded bytearray.
+
+    Each row of the image is encoded separately as required by the DICOM
+    Standard.
+
+    Parameters
+    ----------
+    src : bytes
+        The little-endian ordered data to be encoded, representing a Byte
+        Segment as in the DICOM Standard, Part 5,
+        :dcm:`Annex G.2<part05/sect_G.2.html>`.
+
+    Returns
+    -------
+    bytearray
+        The RLE encoded segment, following the format specified by the DICOM
+        Standard. Odd length encoded segments are padded by a trailing ``0x00``
+        to be even length.
+    """
+    out = bytearray()
+    row_length = kwargs['columns']
+    for idx in range(0, len(src), row_length):
+        out.extend(_encode_row(src[idx:idx + row_length]))
+
+    # Pad odd length data with a trailing 0x00 byte
+    out.extend(b'\x00' * (len(out) % 2))
+
+    return out
+
+
+def _encode_row(src: bytes) -> bytes:
+    """Return `src` as RLE encoded bytes.
+
+    Parameters
+    ----------
+    src : bytes
+        The little-endian ordered data to be encoded.
+
+    Returns
+    -------
+    bytes
+        The RLE encoded row, following the format specified by the DICOM
+        Standard, Part 5, :dcm:`Annex G<part05/chapter_G.html>`
+
+    Notes
+    -----
+    * 2-byte repeat runs are always encoded as Replicate Runs rather than
+      only when not preceeded by a Literal Run as suggested by the Standard.
+    """
+    out = []
+    out_append = out.append
+    out_extend = out.extend
+
+    literal = []
+    for _, group in groupby(src):
+        group = list(group)
+        if len(group) == 1:
+            literal.append(group[0])
+        else:
+            if literal:
+                # Literal runs
+                for ii in range(0, len(literal), 128):
+                    _run = literal[ii:ii + 128]
+                    out_append(len(_run) - 1)
+                    out_extend(_run)
+
+                literal = []
+
+            # Replicate run
+            for ii in range(0, len(group), 128):
+                if len(group[ii:ii + 128]) > 1:
+                    # Replicate run
+                    out_append(257 - len(group[ii:ii + 128]))
```

Comment:
Seems like there is room for optimization here - this and similar loops could be replaced with the integer and remainder from `divisor, remainder = divmod(len(group), 128)` and `extend`s with the (length, val) multiplied by those (or the 1 remainder case). 

But perhaps I am misunderstanding `groupby` or the code somehow...